src/misc/coloring.c

```diff
@@ -129,6 +132,160 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
+    const igraph_vector_int_t *neighbours
+) {
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (VECTOR(*colors)[nbr] == color) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static void igraph_i_dsatur_update_heap(
+    const igraph_adjlist_t *adjlist, igraph_gen2wheap_t *node_degrees_heap,
+    igraph_vector_int_t *neighbours, igraph_vector_int_t *colors,
+    igraph_integer_t color
+) {
+    igraph_gen2wheap_delete_max(node_degrees_heap);
+    dsatur_t color_metadata;
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (!igraph_gen2wheap_has_elem(node_degrees_heap, nbr)) {
+            continue;
+        }
+        color_metadata = *((dsatur_t*)igraph_gen2wheap_get(node_degrees_heap, nbr));
+        if (!is_color_used_by_neighbour(colors, color, igraph_adjlist_get(adjlist, nbr))) {
+            color_metadata.saturation_degree++;
+        }
+        color_metadata.edge_degree--;
+        igraph_gen2wheap_modify(node_degrees_heap, nbr, &color_metadata);
+    }
+}
+
+static igraph_error_t igraph_i_get_first_viable_color(
+    igraph_vector_int_t *viable_colors, igraph_integer_t color_count,
+    igraph_integer_t *result
+) {
+    IGRAPH_CHECK(igraph_vector_int_resize(viable_colors, color_count));
+    igraph_vector_int_sort(viable_colors);
+    igraph_integer_t i = 0;
+    igraph_integer_t col = 0;
+
+    do {
+        while (i < color_count && VECTOR(*viable_colors)[i] == col) {
+            i++;
+        }
+        col++;
+    } while (i < color_count && VECTOR(*viable_colors)[i] == col);
+
+    *result = col;
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_i_dsatur_main_loop(
+    const igraph_t *graph, igraph_vector_int_t *colors,
+    igraph_gen2wheap_t *node_degrees_heap, igraph_adjlist_t *adjlist,
+    igraph_integer_t vc
+) {
+    igraph_integer_t vertices_colored = 0;
+    igraph_vector_int_t viable_colors;
+    igraph_integer_t color;
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&viable_colors, 0);
+
+    while (vertices_colored < vc) {
+        igraph_integer_t node_to_color = igraph_gen2wheap_max_index(node_degrees_heap);
+        igraph_vector_int_t *neighbours = igraph_adjlist_get(adjlist, node_to_color);
+        igraph_integer_t neighbours_count = igraph_vector_int_size(neighbours);
+        igraph_integer_t color_count = 0;
+        IGRAPH_CHECK(igraph_vector_int_resize(&viable_colors, neighbours_count));
+        for (igraph_integer_t neighbor_index = 0; neighbor_index < neighbours_count; neighbor_index++) {
+            igraph_integer_t neighbour = VECTOR(*neighbours)[neighbor_index];
+            if (VECTOR(*colors)[neighbour] != -1) {
```

Comment:
If you are going to skip `-1` value, use `igraph_vector_int_push_back()` to fill the vector. Then you only need to `igraph_vector_int_clear()` the vector at the beginning, but don't need to bother with resizing.